### ISSUE TITLE
Monoid a => Monoid (Behavior t a)?

### ISSUE THREAD

**Issue by ocharles** [HeinrichApfelmus/reactive-banana#78]
This instance would be very useful for folding a collection of behaviors into one - as a generalised `sequenceA`:

``` haskell
let css = fold [ pure defaultStyles
               , fmap (bool mempty (border "1px solid red")) 
                      mouseIsHovering]
               ]
```

Assuming `defaultStyles :: CSS` and `Monoid CSS`.

**Comment by HeinrichApfelmus** [comment]
Hm, I'm not sure about this one. I would write it in a more pedestrian way as

    let css = mkStyle <$> mouseIsHovering
        mkStyle hovering
            =  defaultStyles
            <> (if hovering then mempty else border "1px solid red")

In general, I am a bit wary about `Monoid a => Monoid (Something a)` instances, somehow I always feel like they shut off alternate monoid instances for `Something`.

(The construction would apply to any applicative functor, of course.)

**Comment by ocharles** [comment]
That is true, though I have a hard time agreeing that it is *common* that people would want to combine `Behavior`s out of anything more than a distinguished element and an associative operation - which would be `Monoid a` for `Behavior t a`. Will leave the decision up to you though :)

**Comment by HeinrichApfelmus** [comment]
Well, hypothetically speaking, if `Behavior t a` were not to denote a value that varies over an arbitrarily long time, but a value that varies over a finite time interval, then one monoid operation `(<>)` could be temporal concatenation, i.e. "behavior1 followed by behavior2". Of course, that's not how Behavior currently works, but I still find it interesting.

I think the problem you want to solve arises whenever you have an applicative functor and an infix operation on elements (for example `+`). Then, there is no short (i.e. infix) notation for applying this operation to two arguments. The best you can do at the moment is two arguments

    (+) <$> ax <*> ay

but three arguments seems very awkward.

I will reference issue #34 here, because it touches on a similar problem. I think that if one issue is resolved in a certain way, then the other should probably be resolved in the same way.

**Comment by sjoerdvisscher** [comment]
This should just be

```
let css = under Ap fold [ pure defaultStyles
                        , fmap (bool mempty (border "1px solid red")) 
                               mouseIsHovering]
                        ]
```

with `under` from the newtype package, and `Ap` from the reducers package.

**Comment by ocharles** [comment]
Could be, sure - but I hardly want to force users of my library that uses `reactive-banana` to have to care about this. Ultimately if there is no `Monoid` instance I will just provide my own `mconcat` like function, specialized to `CSS`. That said, I'm not using code in my original example, though I find myself frequently defining the orphan `Monoid a => Monoid (Behavior a)` instance (currently I'm using `WriterT (Behavior [DOMNode]) which requires this instance, or for me to make a `newtype`).

**Comment by ocharles** [comment]
This was fixed in #185
